osu.Game/Skinning/Components/BeatmapAttributeText.cs

```diff
@@ -55,7 +73,35 @@ protected override void LoadComplete()
 
             Attribute.BindValueChanged(_ => updateText());
             Template.BindValueChanged(_ => updateText());
-            beatmap.BindValueChanged(_ => updateText());
+
+            beatmap.BindValueChanged(b =>
+            {
+                difficultyCancellationSource?.Cancel();
+                difficultyCancellationSource = new CancellationTokenSource();
+
+                difficultyBindable?.UnbindAll();
+                difficultyBindable = difficultyCache.GetBindableDifficulty(b.NewValue.BeatmapInfo, difficultyCancellationSource.Token);
+                difficultyBindable.BindValueChanged(d =>
+                {
+                    starDifficulty = d.NewValue;
+                    updateText();
+                });
+
+                updateText();
+            }, true);
+
+            mods.BindValueChanged(m =>
+            {
+                modSettingTracker?.Dispose();
+                modSettingTracker = new ModSettingChangeTracker(m.NewValue)
+                {
+                    SettingChanged = _ => updateText()
+                };
+
+                updateText();
```

Comment:
The thing that bothers me in this implementation is that after all of the work done to ensure that all values are not needlessly precomputed... now every instance of this component is gonna take out a difficulty bindable and a mod setting tracker, and update its text on every star rating and mod change? Even if it'll never use or care about either because it only shows artist name?

And if I were to bet, I'd say those two things alone are probably more of a general resource hog / potential performance issue than the original "weird" precomputed dictionary.

Yes not doing that is probably going to be supremely annoying, unless some other abstraction were to be introduced. And I'm not sure if that should be done, either, but this sort of thing does have me thinking out loud.

As convenient as those two components/mechanisms may be, it feels like we are starting to just spam both of them in every place where we need them, and I'm not sure that is a good thing.

Reply:
> and update its text on every star rating and mod change? Even if it'll never use or care about either because it only shows artist name?

i'd wager the most important parts where you'd want to see this info (song select, gameplay), already have SR & by extension PP calculated by virtue of having gone through song select in the first place - the value is already computed by the wedge and beatmap set panels.

put another way, profiling didn't show any points of concern